### PR TITLE
[Performance][Kernel] Fused_moe Performance Improvement

### PR DESCRIPTION
**Pull request by charlifu**
This PR tries to improve the performance of vllm fuse moe layer by:
- Increase the thread number of `moe_align_block_size_kernel` to warp size.
- Replace `torch.sum` with customized `moe_sum`.
- Move `moe_align_block_size_kernels.cu` to `moe/moe_align_sum_kernels.cu`.

Motivations:
- `moe_align_block_size_kernel` only uses `num_experts` as the thread number, which causes low parallelism when handling big input size, i.e., prefilling.
- At the end of fused_moe, `torch.sum` is used to sum the partial results from all selected experts. But `torch.sum` is not optimized for this use case.

Performance Improvement (Mixtral 8x22b):
moe_sum: **x6** for 8k prefill.
image
moe_align_block_size: **x5** for 8k prefill
image

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by comaniac** [comment]
Thanks for the improvements! Do you have some end-to-end benchmark results to share with?

**Comment by charlifu** [comment]
> Thanks for the improvements! Do you have some end-to-end benchmark results to share with?

Here are the numbers tested on upstream (prefill, tp 2):
| num-prompt | input-len | w/o   | w/    |
|------------|-----------|-------|-------|
| 8          | 8192      | 14.80 | 14.27 |
| 16         | 8192      | 29.84 | 29.21 |
| 32         | 8192      | 59.55 | 57.36 |

But upstream's triton moe kernel is not tuned for rocm, so its huge overhead offsets the performance gain. On rocm fork of vllm, we are able to see around 10% boost on tp8 for Mixtral 8x22b.

**Comment by comaniac** [comment]
Thanks for the numbers. In this case it would be great to come up with tuned configurations (we have them for some NVIDIA GPUs under https://github.com/vllm-project/vllm/tree/main/vllm/model_executor/layers/fused_moe/configs).

Also how does this PR change the performance on NVIDIA GPUs?

**Comment by charlifu** [comment]
> Thanks for the numbers. In this case it would be great to come up with tuned configurations (we have them for some NVIDIA GPUs under https://github.com/vllm-project/vllm/tree/main/vllm/model_executor/layers/fused_moe/configs).
> 
> Also how does this PR change the performance on NVIDIA GPUs?

Nvidia should also benefit from this PR.

**Comment by charlifu** [comment]
@comaniac 
H100 e2e numbers: Mixtral \8x22b, tp8, prefill, 8k input len, 6.25% gain
| num-prompt | input-len | w/o   | w/    |
|------------|-----------|-------|-------|
| 8          | 8192      | 2.00  | 1.88  |
| 16         | 8192      | 3.99  | 3.75  |
| 32         | 8192      | 8.00  | 7.50  |
| 64         | 8192      | 16.01 | 15.04 |
| 128        | 8192      | 32.01 | 30.07 |

MI300X e2e numbers: Mixtral 8x7B, tp8, prefill, 8k input len, 18% gain
| num-prompt | input-len | w/o   | w/    |
|------------|-----------|-------|-------|
| 8          | 8192      | 1.26  | 1.01  |
| 16         | 8192      | 2.53  | 2.04  |
| 32         | 8192      | 5.08  | 4.10  |
| 64         | 8192      | 10.19 | 8.29  |
| 128        | 8192      | 20.53 | 16.78 |

**Comment by comaniac** [comment]
Awesome! Do you want to add AMD tuned configs in this PR or you prefer to have a follow-up PR for it?

**Review by comaniac — APPROVED** [review]
LGTM. I feel we could merge this PR first as it doesn't actually touch the MoE triton kernel.
cc @tlrmchlsmth

**Review by tlrmchlsmth — APPROVED** [review]
Nice! looks good to merge

**Comment by charlifu** [comment]
Looks like it is causing some CI failure on Nvidia side. Looking at it.

**Comment by charlifu** [comment]
The kernel test failure is fixed. I think this PR is ready to be merged. For other two failures, I noticed recently merged PRs failed on those two as well. So I suppose it is not the issue caused by this PR. @tlrmchlsmth @comaniac

**Comment by comaniac** [comment]
Ok I'll ping folks to force merge.
